### ISSUE TITLE
Why doesn't the dataloader of the training data shuffle?

### ISSUE THREAD

**Issue by EIFY** [KellerJordan/modded-nanogpt#174]
Hi, I have run a modified version of modded-nanogpt and noticed that the validation loss oscillates periodically:

<img width="325" height="273" alt="Image" src="https://github.com/user-attachments/assets/c99ab7aa-528f-4cdc-92db-f8c9e3e9789a" />

I went through the code and it seems that the dataloader does not shuffle the training data. I have inspected the current HEAD b7c751d5168879daf34e52faec408907f12bb183 and it still seems to be true for both `train_gpt.py` and `train_gpt_medium.py`. Is there a reason for it? It should be easy and lightweight to at least shuffle at the file level and sequence (within the file) level separately.

**Comment by shenberg** [comment]
Since every training token is only seen once, why is it necessary to shuffle? The LLM regime is usually single-epoch training.

**Comment by EIFY** [comment]
I am about to test shuffled vs. unshuffled training myself but I expect it to matter even with single-epoch training, see e.g.
https://x.com/cis_female/status/2017055613071823318
Think about it this way: If you increasingly put the tokens about math at the beginning, the model is increasingly doing continual learning as a math model learning other topics, with worse and worse expected catastrophic forgetting

**Comment by shenberg** [comment]
Interesting. I was assuming that FineWeb documents were ordered by their unique id per drop, but that's unclear. The docs say that [FineWeb-10B is a random subset of FineWeb](https://huggingface.co/datasets/HuggingFaceFW/fineweb#smaller-sample-versions) so should ideally be roughly shuffled anyhow. "Should" is doing a lot of work there, though. I'm interested in the result.

**Comment by trianxy** [comment]
> Is there a reason for it? It should be easy and lightweight to at least shuffle at the file level and sequence (within the file) level separately.

As @shenberg mentioned, I think it's because the underlying training tokens are already shuffled.

Moreover, optimizing the pipeline by changing the order of training tokens should be possible, but is out of scope for this competition (but not for the competition in https://github.com/mlfoundations/dclm !): The [rules](https://github.com/KellerJordan/modded-nanogpt?tab=readme-ov-file#rules) say to `Not modify the train or validation data pipelines.` The focus in this competition is on the model architecture and its implementation.

**Comment by EIFY** [comment]
@trianxy Circling back to close the loop: I can't find an official confirmation but it really doesn't seem that these 10B / 100B subsets are shuffled. HuggingFace hosts official shuffled version of Fineweb 100BT / FineWeb-Edu 100BT separately. If they were already shuffled it wouldn't be necessary:

https://huggingface.co/datasets/HuggingFaceFW/fineweb_100BT-shuffled
https://huggingface.co/datasets/HuggingFaceFW/fineweb_edu_100BT-shuffled

That said, rules are rules. I guess the unshuffled tokens are just the original sin we need to live with. I am closing the issue and editing my original post for public awareness.

**Comment by EIFY** [comment]
Addendum: There are other changes e.g. the batch size is 4x as before so it's not exactly apples-to-apples, but switching to [fineweb_edu_100BT-shuffled](https://huggingface.co/datasets/HuggingFaceFW/fineweb_edu_100BT-shuffled) eliminated the validation loss oscillations as expected:
<img width="517" height="272" alt="Image" src="https://github.com/user-attachments/assets/059dd48c-9a75-4e51-a996-78bc2c789c67" />
